File: packages/js/core/src/types/MaybeAsync.ts
export type MaybeAsync<T> = Promise<T> | T;

export const isPromise = <T extends unknown>(
  test?: MaybeAsync<T>
): test is Promise<T> =>
  !!test && typeof (test as Promise<T>).then === "function";

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export const executeMaybeAsyncFunction = async <T extends unknown>(
  func: (...args: unknown[]) => Promise<T> | T,
  ...args: unknown[]
): Promise<T> => {
  let result = func(...args);
  if (isPromise(result)) {
    result = await result;
  }
  return result;
};

File: packages/py/core/src/types/maybe_async.py
from __future__ import annotations
import inspect
from typing import Awaitable, Optional, Union, Any, Callable


def is_promise(test: Optional[Union[Awaitable, Any]] = None) -> bool:
    return test and inspect.iscoroutine(test)


async def execute_maybe_async_function(func: Callable, *args: Any) -> Any:
    if args:
        result = func(args)
    else:
        result = func()
    if is_promise(result):
        result = await result
    return result
